File: python-engine/services/benchmark_service.py
"""
Benchmark service for orchestrating formula calculations.

This module provides the BenchmarkService class that coordinates
the complete benchmark workflow: reading data, calculating formulas,
saving results, and logging execution times.
"""

import logging
import time
from typing import List, Optional

import sys
import os
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from engines.base_engine import IFormulaEngine, BenchmarkResult, FormulaPerformance
from repositories.postgres_repository import PostgresRepository
from models.formula import Formula
from models.data_record import DataRecord
from models.calculation_result import CalculationResult


# Configure module-level logger
logger = logging.getLogger(__name__)


class BenchmarkService:
    """
    Service for orchestrating benchmark calculations.
    
    This service coordinates the entire benchmark workflow:
    1. Fetches formulas from the database
    2. Iterates through data records in batches (100,000 per batch)
    3. Calculates each formula using the provided engine
    4. Saves results to t_results
    5. Logs execution times to t_log
    
    Uses batch processing to efficiently handle 1 million records
    without excessive memory consumption.
    """
    
    # Batch size for processing records (100,000 records per batch)
    BATCH_SIZE = 100_000
    
    def __init__(
        self, 
        engine: IFormulaEngine, 
        repository: PostgresRepository,
        clear_previous_results: bool = True
    ):
        """
        Initialize the benchmark service.
        
        Args:
            engine: The formula calculation engine to use
            repository: Database repository for data access
            clear_previous_results: Whether to clear previous results before running
        """
        self._engine = engine
        self._repository = repository
        self._clear_previous = clear_previous_results
    
    @property
    def method_name(self) -> str:
        """Get the calculation method name from the engine."""
        return self._engine.name
    
    async def run_benchmark(self) -> BenchmarkResult:
        """
        Run the complete benchmark for all formulas.
        
        This method:
        1. Clears previous results for this method (if configured)
        2. Fetches all formulas from t_targil
        3. For each formula, processes all data records in batches
        4. Saves results and logs timing for each formula
        
        Returns:
            BenchmarkResult with timing metrics for all formulas
        """
        overall_start = time.perf_counter()
        formula_performances: List[FormulaPerformance] = []
        
        logger.info(f"Starting benchmark for method: {self.method_name}")
        
        # Clear previous results if configured
        if self._clear_previous:
            await self._repository.clear_method_results(self.method_name)
            await self._repository.clear_method_logs(self.method_name)
        
        # Fetch all formulas
        formulas = await self._repository.get_all_formulas()
        logger.info(f"Processing {len(formulas)} formulas")
        
        # Get total record count for logging
        total_records = await self._repository.get_data_record_count()
        logger.info(f"Total data records to process: {total_records:,}")
        
        # Process each formula
        for formula_idx, formula in enumerate(formulas, 1):
            performance = await self._process_formula(
                formula, 
                formula_idx, 
                len(formulas),
                total_records
            )
            formula_performances.append(performance)
        
        overall_end = time.perf_counter()
        total_time = overall_end - overall_start
        
        logger.info(
            f"Benchmark complete for {self.method_name}: "
            f"{len(formulas)} formulas, {total_records:,} records, "
            f"total time: {total_time:.2f}s"
        )
        
        return BenchmarkResult(
            method=self.method_name,
            total_time=total_time,
            formula_results=formula_performances
        )
    
    async def _process_formula(
        self, 
        formula: Formula, 
        formula_idx: int,
        total_formulas: int,
        total_records: int
    ) -> FormulaPerformance:
        """
        Process a single formula across all data records.
        
        Args:
            formula: The formula to calculate
            formula_idx: Current formula number (1-indexed)
            total_formulas: Total number of formulas
            total_records: Total number of data records
            
        Returns:
            FormulaPerformance with timing metrics
        """
        formula_start = time.perf_counter()
        records_processed = 0
        batch_num = 0
        
        logger.info(
            f"[{formula_idx}/{total_formulas}] Processing formula {formula.targil_id}: "
            f"'{formula.targil}'"
            + (f" (conditional: {formula.tnai})" if formula.is_conditional else "")
        )
        
        # Process data in batches
        async for batch in self._repository.iterate_data_batches():
            batch_num += 1
            batch_start = time.perf_counter()
            
            # Calculate formula for this batch
            results = await self._engine.calculate_formula(formula, batch)
            
            # Save results
            await self._repository.save_results_batch(results)
            
            records_processed += len(batch)
            batch_time = time.perf_counter() - batch_start
            
            # Progress logging every batch
            progress_pct = (records_processed / total_records) * 100
            logger.debug(
                f"  Batch {batch_num}: {len(batch):,} records, "
                f"{batch_time:.2f}s, progress: {progress_pct:.1f}%"
            )
        
        formula_end = time.perf_counter()
        execution_time = formula_end - formula_start
        
        # Log timing to database
        await self._repository.save_log_entry(
            targil_id=formula.targil_id,
            method=self.method_name,
            run_time=execution_time,
            records_processed=records_processed
        )
        
        logger.info(
            f"  Completed formula {formula.targil_id}: "
            f"{records_processed:,} records in {execution_time:.2f}s "
            f"({records_processed / execution_time:.0f} records/sec)"
        )
        
        return FormulaPerformance(
            targil_id=formula.targil_id,
            formula=formula.targil,
            execution_time=execution_time,
            records_processed=records_processed
        )
    
    async def run_single_formula(
        self, 
        targil_id: int,
        save_results: bool = True
    ) -> Optional[FormulaPerformance]:
        """
        Run benchmark for a single formula.
        
        Useful for testing or re-running a specific formula.
        
        Args:
            targil_id: The formula ID to process
            save_results: Whether to save results to database
            
        Returns:
            FormulaPerformance with timing metrics, or None if formula not found
        """
        formulas = await self._repository.get_all_formulas()
        formula = next((f for f in formulas if f.targil_id == targil_id), None)
        
        if formula is None:
            logger.error(f"Formula {targil_id} not found")
            return None
        
        total_records = await self._repository.get_data_record_count()
        
        return await self._process_formula(
            formula,
            formula_idx=1,
            total_formulas=1,
            total_records=total_records
        )
    
    async def verify_results(
        self, 
        sample_size: int = 100
    ) -> dict:
        """
        Verify a sample of results matches expected calculations.
        
        Args:
            sample_size: Number of data records to verify
            
        Returns:
            Dict with verification status and any discrepancies
        """
        formulas = await self._repository.get_all_formulas()
        discrepancies = []
        
        for formula in formulas:
            results = await self._repository.get_results_for_verification(
                formula.targil_id, 
                limit=sample_size
            )
            
            # Group results by data_id
            results_by_data: dict = {}
            for r in results:
                data_id = r['data_id']
                if data_id not in results_by_data:
                    results_by_data[data_id] = {}
                results_by_data[data_id][r['method']] = r['result']
            
            # Check consistency (this method's results exist)
            for data_id, methods in results_by_data.items():
                if self.method_name not in methods:
                    discrepancies.append({
                        'type': 'missing_result',
                        'data_id': data_id,
                        'targil_id': formula.targil_id,
                        'method': self.method_name
                    })
        
        return {
            'verified': len(discrepancies) == 0,
            'sample_size': sample_size,
            'formulas_checked': len(formulas),
            'discrepancies': discrepancies
        }


async def create_benchmark_service(
    connection_string: str,
    engine: Optional[IFormulaEngine] = None,
    clear_previous: bool = True
) -> BenchmarkService:
    """
    Factory function to create and initialize a BenchmarkService.
    
    Args:
        connection_string: PostgreSQL connection string
        engine: Optional formula engine (creates PythonFormulaEngine if None)
        clear_previous: Whether to clear previous results
        
    Returns:
        Initialized BenchmarkService
    """
    # Import here to avoid circular imports
    from engines.eval_engine import PythonFormulaEngine
    
    # Create repository and connect
    repository = PostgresRepository(
        connection_string, 
        batch_size=BenchmarkService.BATCH_SIZE
    )
    await repository.connect()
    
    # Create engine if not provided
    if engine is None:
        engine = PythonFormulaEngine()
        await engine.initialize(connection_string)
    
    return BenchmarkService(
        engine=engine,
        repository=repository,
        clear_previous_results=clear_previous
    )

File: dotnet-engine/FormulaEngine/Services/BenchmarkService.cs
using System.Diagnostics;
using FormulaEngine.Engines;
using FormulaEngine.Models;
using FormulaEngine.Repositories;

namespace FormulaEngine.Services;

/// <summary>
/// Service that orchestrates the formula benchmark execution.
/// Coordinates data loading, formula calculation, and result persistence.
/// </summary>
public class BenchmarkService
{
    private readonly IDataRepository _repository;
    private readonly IFormulaEngine _engine;

    public BenchmarkService(IDataRepository repository, IFormulaEngine engine)
    {
        _repository = repository;
        _engine = engine;
    }

    /// <summary>
    /// Run the complete benchmark for all formulas.
    /// Measures execution time per formula and saves results to the database.
    /// </summary>
    public async Task<BenchmarkSummary> RunBenchmarkAsync()
    {
        var summary = new BenchmarkSummary
        {
            Method = _engine.Name,
            StartTime = DateTime.Now
        };

        var totalStopwatch = Stopwatch.StartNew();

        try
        {
            Console.WriteLine($"\n{'=',-60}");
            Console.WriteLine($"Starting benchmark for method: {_engine.Name}");
            Console.WriteLine($"{'=',-60}\n");

            // Clear previous results for this method
            Console.WriteLine("Clearing previous results...");
            await _repository.ClearMethodResultsAsync(_engine.Name);
            await _repository.ClearMethodLogsAsync(_engine.Name);

            // Load data records
            Console.WriteLine("Loading data records from database...");
            var dataLoadStopwatch = Stopwatch.StartNew();
            var dataRecords = (await _repository.GetAllDataRecordsAsync()).ToList();
            dataLoadStopwatch.Stop();
            Console.WriteLine($"Loaded {dataRecords.Count:N0} data records in {dataLoadStopwatch.Elapsed.TotalSeconds:F2}s\n");

            summary.TotalRecords = dataRecords.Count;

            // Load formulas
            Console.WriteLine("Loading formulas from database...");
            var formulas = (await _repository.GetAllFormulasAsync()).ToList();
            Console.WriteLine($"Found {formulas.Count} formulas to process\n");

            summary.TotalFormulas = formulas.Count;

            // Process each formula
            foreach (var formula in formulas)
            {
                var formulaResult = await ProcessFormulaAsync(formula, dataRecords);
                summary.FormulaResults.Add(formulaResult);
            }

            totalStopwatch.Stop();
            summary.TotalTime = totalStopwatch.Elapsed.TotalSeconds;
            summary.EndTime = DateTime.Now;

            // Print summary
            PrintSummary(summary);

            return summary;
        }
        catch (Exception ex)
        {
            totalStopwatch.Stop();
            Console.WriteLine($"\nBenchmark failed with error: {ex.Message}");
            throw;
        }
    }

    /// <summary>
    /// Process a single formula: calculate results for all data records,
    /// measure execution time, and save to database.
    /// </summary>
    private async Task<FormulaResult> ProcessFormulaAsync(Formula formula, List<DataRecord> dataRecords)
    {
        var formulaDisplay = GetFormulaDisplay(formula);
        Console.WriteLine($"Processing formula {formula.TargilId}: {formulaDisplay}");

        var stopwatch = Stopwatch.StartNew();

        // Calculate results for all data records
        var results = _engine.CalculateFormula(formula, dataRecords);

        stopwatch.Stop();
        var executionTime = stopwatch.Elapsed.TotalSeconds;

        // Save results to database
        Console.Write($"  Calculated {results.Length:N0} results in {executionTime:F3}s - Saving...");
        await _repository.SaveResultsBatchAsync(results);
        Console.WriteLine(" Done");

        // Save timing log
        await _repository.SaveLogEntryAsync(formula.TargilId, _engine.Name, executionTime);

        // Count valid/null results
        var validResults = results.Count(r => r.Result.HasValue);
        var nullResults = results.Length - validResults;

        Console.WriteLine($"  Valid results: {validResults:N0}, Null results: {nullResults:N0}\n");

        return new FormulaResult
        {
            TargilId = formula.TargilId,
            Formula = formulaDisplay,
            ExecutionTime = executionTime,
            RecordsProcessed = results.Length,
            ValidResults = validResults,
            NullResults = nullResults
        };
    }

    /// <summary>
    /// Get a display string for the formula, including condition if present.
    /// </summary>
    private static string GetFormulaDisplay(Formula formula)
    {
        if (!string.IsNullOrEmpty(formula.Tnai))
        {
            var falseFormula = formula.TargilFalse ?? "null";
            return $"if({formula.Tnai}, {formula.Targil}, {falseFormula})";
        }
        return formula.Targil;
    }

    /// <summary>
    /// Print the benchmark summary to the console.
    /// </summary>
    private static void PrintSummary(BenchmarkSummary summary)
    {
        Console.WriteLine($"\n{'=',-60}");
        Console.WriteLine("BENCHMARK SUMMARY");
        Console.WriteLine($"{'=',-60}");
        Console.WriteLine($"Method:           {summary.Method}");
        Console.WriteLine($"Total Records:    {summary.TotalRecords:N0}");
        Console.WriteLine($"Total Formulas:   {summary.TotalFormulas}");
        Console.WriteLine($"Total Time:       {summary.TotalTime:F2} seconds");
        Console.WriteLine($"Start Time:       {summary.StartTime:yyyy-MM-dd HH:mm:ss}");
        Console.WriteLine($"End Time:         {summary.EndTime:yyyy-MM-dd HH:mm:ss}");
        Console.WriteLine($"\nPer-Formula Results:");
        Console.WriteLine($"{"-",-60}");
        Console.WriteLine($"{"ID",-4} {"Formula",-35} {"Time (s)",-10} {"Valid",-12}");
        Console.WriteLine($"{"-",-60}");

        foreach (var result in summary.FormulaResults)
        {
            var formulaShort = result.Formula.Length > 32 
                ? result.Formula.Substring(0, 29) + "..." 
                : result.Formula;
            Console.WriteLine($"{result.TargilId,-4} {formulaShort,-35} {result.ExecutionTime,-10:F3} {result.ValidResults:N0}");
        }

        Console.WriteLine($"{'=',-60}\n");
    }
}

/// <summary>
/// Summary of a complete benchmark run.
/// </summary>
public class BenchmarkSummary
{
    public string Method { get; set; } = string.Empty;
    public int TotalRecords { get; set; }
    public int TotalFormulas { get; set; }
    public double TotalTime { get; set; }
    public DateTime StartTime { get; set; }
    public DateTime EndTime { get; set; }
    public List<FormulaResult> FormulaResults { get; } = new();
}

/// <summary>
/// Result of processing a single formula.
/// </summary>
public class FormulaResult
{
    public int TargilId { get; set; }
    public string Formula { get; set; } = string.Empty;
    public double ExecutionTime { get; set; }
    public int RecordsProcessed { get; set; }
    public int ValidResults { get; set; }
    public int NullResults { get; set; }
}
